script/dump_data.rb

```diff
@@ -0,0 +1,648 @@
+# Returns tables in sorted order
+def get_tables_to_process
+  tables = ActiveRecord::Base.connection.tables.sort
+
+  # Filter out internal tables
+  tables.reject! do |table|
+    SKIP_TABLES.include?(table) || table =~ /^(MLOG\$|RUPD\$|BIN\$)/
+  end
+
+  tables
+end
+
+def export_data
+  tables = get_tables_to_process
+
+  # Tables with foreign key constraints that should be imported last
+  # Based on foreign keys defined in db/oracle_schema.rb
+  tables_with_fk_dependencies = [
+    'api_docs_services',      # FK to services
+    'payment_details',        # FK to accounts
+    'policies',               # FK to accounts
+    'provided_access_tokens', # FK to users
+    'proxy_configs',          # FK to proxies, users
+    'sso_authorizations'      # FK to authentication_providers, users
+  ]
+
+  # Move tables with foreign keys to the end in dependency order
+  tables_with_fk_dependencies.each do |table|
+    tables << table if tables.delete(table)
+  end
+
+  STDERR.puts "Exporting data from #{tables.size} tables..."
+  STDERR.puts "=" * 80
+
+  exported_count = 0
+  total_records = 0
+
+  tables.each do |table|
+    begin
+      # Try to find model for this table
+      model = find_model_for_table(table)
+
+      # Get total count first
+      record_count = if model
+        model.count
+      else
+        count_sql = "SELECT COUNT(*) FROM #{ActiveRecord::Base.connection.quote_table_name(table)}"
+        count_result = ActiveRecord::Base.connection.exec_query(count_sql)
+        count_result.rows.first.first
+      end
+
+      if record_count == 0
+        STDERR.puts "#{table}: 0 records (empty)"
+        # Output empty batch to ensure table gets truncated on import
+        puts JSON.generate({ table: table, data: [] })
+        exported_count += 1
+        next
+      end
+
+      STDERR.print "#{table}: exporting #{record_count} records..."
+
+      batch_count = 0
+
+      if model
+        # Check if table has composite primary key
+        has_composite_key = model.primary_key.nil? || model.primary_key.is_a?(Array)
+
+        # Get valid column names to filter out virtual attributes (like tag_list)
+        valid_columns = model.column_names
+
+        if has_composite_key
+          # For composite primary keys, fetch all at once using the model
+          # find_each requires ORDER BY which doesn't work with composite keys
+          # Use model.all to get proper type serialization
+          # Filter to only include actual database columns (exclude virtual attributes)
+          all_records = model.all.map { |r| r.attributes.slice(*valid_columns) }
+
+          all_records.each_slice(BATCH_SIZE) do |batch|
+            puts JSON.generate({ table: table, data: batch })
+            batch_count += 1
+            total_records += batch.size
+          end
+        else
+          # Identify serialized columns
+          serialized_columns = []
+          if model.respond_to?(:serialized_attributes)
+            # Rails < 7.1
```

Comment:
Does 2.16 include the Rails 7.1 upgrade? If so, we might not need this conditional.